p2p/transport.go

```diff
@@ -537,6 +537,18 @@ func handshake(
 		}
 	}
 
+	addr, err := peerNodeInfo.NetAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the node is reporting a non-routable address then we use the remote IP from the connection as the ListenAddr.
+	// While this doesn't prevent adding bogus IPs, it does help ensure we use routable IPs for inbound node info.
+	// Note: If the remote address is not routable (i.e. private network) then ListenAddr will still be set to this private address.
+	if !addr.Routable() {
+		peerNodeInfo.ListenAddr = c.RemoteAddr().String()
```

Comment:
Unfortunately, `RemoteAddr()` will include the arbitrarily chosen outbound port number of the remote IP, not the port number which the peer is (possibly) listening on. I believe this will wind up in the address book, and may be gossiped to other peers who will not be able to connect.

As a best effort we may discard the port number and fall back to the default Tendermint port number, but even this is somewhat brittle.